### PR TITLE
Deduplicate unsatisfied trait bounds

### LINKED ISSUE #35677: the method exists but the following trait bounds were not satisfied repeats the bounds multiple times
``` rust
use std::collections::HashMap;
fn intersect_map<K, V>(this: &mut HashMap<K, V>, other: HashMap<K, V>) -> bool {
    this.drain()
}
```

reports

```
note: the method `drain` exists but the following trait bounds were not satisfied: `K : std::cmp::Eq`, `K : std::hash::Hash`, `K : std::cmp::Eq`, `K : std::hash::Hash`
```

Note how it reports about `K` missing necessary bounds twice.

### PR DESCRIPTION
**Pull request by estebank**
Fix #35677.

### CODE REVIEW

File: src/librustc_typeck/check/method/suggest.rs [hunk 1]
```diff
@@ -424,10 +424,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
 
                 if !unsatisfied_predicates.is_empty() {
-                    let bound_list = unsatisfied_predicates.iter()
+                    let mut bound_list = unsatisfied_predicates.iter()
                         .map(|p| format!("`{} : {}`", p.self_ty(), p))
-                        .collect::<Vec<_>>()
-                        .join("\n");
+                        .collect::<Vec<_>>();
+                    bound_list.sort();
+                    bound_list.dedup_by(|a, b| a == b);  // #35677
```
Comment (oli-obk) [inline src/librustc_typeck/check/method/suggest.rs]: isn't this equivalent to 

```suggestion
                    bound_list.dedup();  // #35677
```

**Review by oli-obk — CHANGES_REQUESTED** [review]
just a nit.

File: src/librustc_typeck/check/method/suggest.rs (hunk 1 above)
Comment (estebank) [reply src/librustc_typeck/check/method/suggest.rs]: It is, didn't see it when searching for it in the docs, it turns out it was just out of order :)

**Comment by alexreg** [comment]
Is this related to my ongoing PR https://github.com/rust-lang/rust/pull/55994?

**Comment by oli-obk** [comment]
I don't think so, this PR relates to collecting duplicates while producing a diagnostic. This can happen if bounds exist on an `impl` block and on a function

**Review by oli-obk — APPROVED** [review]

**Pull request merged**